doc/spec.md

```diff
@@ -2348,7 +2355,62 @@ foo({ id: 1234, name: false });    // Error, name of wrong type
+
+*S* is ***comparable to*** a type *T*, and *T* is ***comparable from*** *S*, if one of the following is true:
+
+* *S* and *T* are identical types.
+* *S* or *T* is the Any type.
+* *S* or *T* is an enum type and the other is the primitive type Number.
+* *S* is a literal type and *T* is the base primitive type of S.
+* *S* is a union type and some constituent type of *S* is comparable to *T*.
+* *S* is an intersection type and at least one constituent type of *S* is comparable to *T*.
+* *T* is a union type and *S* is comparable to at least one constituent type of *T*.
+* *T* is an intersection type and *S* is comparable to each constituent type of *T*.
+* *S* is a type parameter and the constraint of *S* is comparable to *T*.
+* *S* is an object type, an intersection type, an enum type, or the Number, Boolean, or String primitive type, *T* is an object type, and for each member *M* in *T*, one of the following is true:
+  * *M* is a property and *S* has an apparent property *N* where
+    * *M* and *N* have the same name,
+    * the type of *N* is comparable to that of *M*,
+    * *M* and *N* are both public, *M* and *N* are both private and originate in the same declaration, *M* and *N* are both protected and originate in the same declaration, or *M* is protected and *N* is declared in a class derived from the class in which *M* is declared.
+  * *M* is an optional property and *S* has no apparent property of the same name as *M*.
+  * *M* is a non-specialized call or construct signature and *S* has an apparent call or construct signature *N* where, when *M* and *N* are instantiated using type Any as the type argument for all type parameters declared by *M* and *N* (if any),
+    * the signatures are of the same kind (call or construct),
+    * *M* has a rest parameter or the number of non-optional parameters in *N* is less than or equal to the total number of parameters in *M*,
+    * for parameter positions that are present in both signatures, each parameter type in *N* is comparable to or from the corresponding parameter type in *M*, and
+    * the result type of *M* is Void, or the result type of *N* is comparable to that of *M*.
+  * *M* is a string index signature of type *U*, and *U* is the Any type or *S* has an apparent string index signature of a type that is comparable to *U*.
+  * *M* is a numeric index signature of type *U*, and *U* is the Any type or *S* has an apparent string or numeric index signature of a type that is comparable to *U*.
+
+When comparing call or construct signatures, parameter names are ignored and rest parameters correspond to an unbounded expansion of optional parameters of the rest parameter element type.
+
+Note that specialized call and construct signatures (section [3.9.2.4](#3.9.2.4)) are not significant when determining assignment compatibility.
```

Comment:
Specialized signatures are not a thing anymore, I think. Though we should check whether the rules for validating overload sets are up to date.